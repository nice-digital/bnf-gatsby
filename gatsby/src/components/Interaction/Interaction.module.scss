@use '@nice-digital/nds-core/scss/core';

.messageList {
  list-style: none;
  margin: 0;
}

.interactantTitle {
  @include core.nds-h5();
}

.message {
  margin-bottom: core.rem(core.$nds-spacing-medium);
}

.message + .message {
  border-top: 2px dashed core.$nds-colour-nice-grey-1;
  padding-top: core.rem(core.$nds-spacing-medium);

  p:first-of-type {
    margin-top: 0;
  }
}

.severeMessage {
  border-left: core.rem(core.$nds-spacing-small) solid core.$nds-colour-error;
  padding-left: core.rem(core.$nds-spacing-medium);
}

.severeText {
  color: core.$nds-colour-error;
  display: inline-block;
  margin-right: core.rem(core.$nds-spacing-small);
}

.supplementaryInfo {
  margin-top: core.rem(core.$nds-spacing-medium);

  dt,
  dd {
    display: inline;
    margin: 0 core.rem(core.$nds-spacing-small) 0 0;
  }

  dd + dt {
    margin-left: core.rem(core.$nds-spacing-small);
  }
}
