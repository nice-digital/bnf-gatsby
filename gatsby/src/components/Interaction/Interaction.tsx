import { Link } from "gatsby";
import React from "react";

import styles from "./Interaction.module.scss";

export type InteractionProps = {
	interactant: {
		title: string;
		drug: {
			slug: string;
		};
	};
	messages: {
		additiveEffect: boolean;
		evidence: string | null;
		message: string;
		severity: string;
		severityOrder: number;
	}[];
};

export const Interaction: React.FC<InteractionProps> = ({
	interactant,
	messages,
}: InteractionProps) => {
	const SEVERE_INTERACTION = 4;

	return (
		<>
			<h3 className={styles.interactantTitle}>
				{interactant.drug?.slug ? (
					<Link to={`/drugs/${interactant.drug.slug}/`}>
						{interactant.title}
					</Link>
				) : (
					interactant.title
				)}
			</h3>
			<ul className={styles.messageList}>
				{messages.map(
					({ message, severity, severityOrder, evidence }, messageIndex) => {
						const isSevereInteraction = severityOrder >= SEVERE_INTERACTION,
							isNormalInteraction = severity === "Normal";
						return (
							<li key={messageIndex} className={styles.message}>
								<div
									className={isSevereInteraction ? styles.severeMessage : ""}
								>
									<div dangerouslySetInnerHTML={{ __html: message }}></div>
									{(evidence !== null || !isNormalInteraction) && (
										<dl className={styles.supplementaryInfo}>
											{!isNormalInteraction && (
												<>
													<dt>Severity:</dt>
													<dd>{severity}</dd>
												</>
											)}
											{evidence !== null && (
												<>
													<dt>Evidence:</dt>
													<dd>{evidence}</dd>
												</>
											)}
										</dl>
									)}
								</div>
							</li>
						);
					}
				)}
			</ul>
		</>
	);
};
