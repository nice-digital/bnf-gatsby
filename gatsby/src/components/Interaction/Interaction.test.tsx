import { render, screen, within } from "@testing-library/react";
import React from "react";

import { Interaction, type InteractionProps } from "./Interaction";

const props: InteractionProps = {
	interactant: {
		title: "Test title",
		drug: {
			slug: "test-title",
		},
	},
	messages: [
		{
			additiveEffect: false,
			evidence: "Test evidence",
			message: "Lorem ipsum dolor sit amet",
			severity: "Severe",
			severityOrder: 4,
		},
		{
			additiveEffect: true,
			evidence: "More test evidence",
			message:
				"<p>Fusce in justo tincidunt, blandit est at, feugiat tellus.</p>",
			severity: "Normal",
			severityOrder: 1,
		},
	],
};

describe("Interaction", () => {
	it("Should match the snapshot", () => {
		render(<Interaction {...props} />);
		expect(document.body).toMatchSnapshot();
	});

	it("Should render the interactant's title as an h3", () => {
		render(<Interaction {...props} />);
		const heading = screen.getByRole("heading", {
			name: props.interactant.title,
			level: 3,
		});
		expect(heading).toBeInTheDocument();
	});

	it("Should render a message for each entry in the messages array", () => {
		render(<Interaction {...props} />);
		const messageListItems = screen.getAllByRole("listitem");
		expect(messageListItems).toHaveLength(props.messages.length);
	});

	it("Should render an appropriate class and message when the interaction is severe", () => {
		render(<Interaction {...props} />);
		const severeListItem = screen.getAllByRole("listitem")[0];
		// eslint-disable-next-line testing-library/no-node-access
		expect(severeListItem.firstChild).toHaveClass("severeMessage");

		const severeParagraph = within(severeListItem).getByText("Severe");
		expect(severeParagraph).toBeInTheDocument;
	});
});
