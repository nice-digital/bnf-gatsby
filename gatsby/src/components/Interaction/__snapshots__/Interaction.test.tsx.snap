// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interaction Should match the snapshot 1`] = `
<body>
  <div>
    <h3
      class="interactantTitle"
    >
      <a
        href="/drugs/test-title/"
      >
        Test title
      </a>
    </h3>
    <ul
      class="messageList"
    >
      <li
        class="message"
      >
        <div
          class="severeMessage"
        >
          <div>
            Lorem ipsum dolor sit amet
          </div>
          <dl
            class="supplementaryInfo"
          >
            <div
              class="severeText"
            >
              <dt>
                Severity:
              </dt>
              <dd>
                Severe
              </dd>
            </div>
            <dt>
              Evidence:
            </dt>
            <dd>
              Test evidence
            </dd>
          </dl>
        </div>
      </li>
      <li
        class="message"
      >
        <div
          class=""
        >
          <div>
            <p>
              Fusce in justo tincidunt, blandit est at, feugiat tellus.
            </p>
          </div>
          <dl
            class="supplementaryInfo"
          >
            <dt>
              Evidence:
            </dt>
            <dd>
              More test evidence
            </dd>
          </dl>
        </div>
      </li>
    </ul>
  </div>
</body>
`;
